cdc/log.cc

```diff
@@ -974,11 +976,12 @@ cdc::cdc_service::impl::augment_mutation_call(lowres_clock::time_point timeout,
         return make_ready_future<std::tuple<std::vector<mutation>, cdc::result_callback>>(std::make_tuple(std::move(mutations), result_callback{}));
     }
 
+    tracing::trace(tr_state, "CDC: Started generating mutations for log rows");
     mutations.reserve(2 * mutations.size());
 
     return do_with(std::move(mutations), service::query_state(service::client_state::for_internal_calls(), empty_service_permit()),
-            [this, timeout, i] (std::vector<mutation>& mutations, service::query_state& qs) {
-        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs] (int idx) {
+            [this, timeout, i, tr_state = std::move(tr_state)] (std::vector<mutation>& mutations, service::query_state& qs) {
+        return transform_mutations(mutations, 1, [this, &mutations, timeout, &qs, tr_state = tr_state] (int idx) mutable {
```

Comment:
this `tr_state=tr_state` makes no sense to me.
The problem you tried to describe earlier is simply because of the missing `mutable` in the external lambda `tr_state` is already const here.

Reply:
This assignment removes constness of `tr_state`. The variable accessible from the inside will be modifiable.

You can verify that `tr_state` is modifiable inside this lambda - If you place `tr_state = nullptr;` on the next line, it will work.

Reply:
It's not removing the constness from `tr_state` - it declares a variable with the name `tr_state` in the context of the internal lambda which initialized using a copy-constructor with a value of a `tr_state` from the external lambda.

The `tr_state` in the internal lambda is non-const because you used `mutable` in the declaration of the internal lambda.

This code is functionally correct (I thought it wasn't but now I see that it is) but it's very cryptic and hacky.

I suggest you either put a very fat comment that explains that variable copy trick, or you can put `mutable` in the external lambda declaration and thereby avoid the need for the "trick".

I'd personally prefer the later.

Reply:
> This code is functionally correct (I thought it wasn't but now I see that it is) but it's very cryptic and hacky.
> I suggest you either put a very fat comment that explains that variable copy trick, or you can put mutable in the external lambda declaration and thereby avoid the need for the "trick".

With the "trick" we have better const correctness (because we don't have to make every lambda on the path to this mutable), but I guess that if it comes as too surprising and cryptic, then the benefit is lost. I will change to `mutable` in a follow up PR.